{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json, csv\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "\n",
    "#--- Variables: Test File Locations ---\n",
    "test_fasta_file = \"data\\\\test\\\\test_proteins.fasta\"\n",
    "test_json_file = \"data\\\\test\\\\test_proteins.json\"\n",
    "# motif_file = \"\"\n",
    "test_ifeature_file = \"data\\\\test\\\\test_iFeature.json\"\n",
    "test_names_file = \"data\\\\test\\\\test_names.tsv\"\n",
    "test_output_file = \"data\\\\test\\\\test_protein_information.json\"\n",
    "\n",
    "#--- Variables: Training File Locations ---\n",
    "train_protein_file = \"data\\\\train\\\\train_proteins.fasta\"\n",
    "train_uniprot_file = \"data\\\\train\\\\train_proteins.json\"\n",
    "train_ifeature_file = \"data\\\\train\\\\train_iFeature.json\"\n",
    "train_output_file = \"data\\\\train\\\\train_protein_information.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#--- Functions: Assign Annotations (UniProt) ---\n",
    "def assign_annotations(json_file, proteins):\n",
    "    # Load JSON data from UniProt file\n",
    "    with open(json_file, 'r') as file:\n",
    "        data = json.load(file)\n",
    "        results = data[\"results\"]\n",
    "    \n",
    "    for result in results:\n",
    "        id = result[\"primaryAccession\"]\n",
    "        if \"uniProtKBCrossReferences\" in result: proteins[id][\"go_terms\"] = get_go_terms(result)\n",
    "        if \"comments\" in result: proteins[id][\"subcell_locations\"] = get_subcellular_locations(result)\n",
    "        if \"features\" in result: proteins[id][\"transmembrane\"] = get_transmembrane(result)\n",
    "        if \"keywords\" in result: proteins[id][\"binding_preference\"] = get_binding_preference(result)\n",
    "\n",
    "def get_go_terms(protein_info):\n",
    "    go_terms = []\n",
    "    references = protein_info[\"uniProtKBCrossReferences\"]\n",
    "    for reference in references:\n",
    "        if reference[\"database\"] == \"GO\":\n",
    "            go_id = reference[\"id\"]\n",
    "            go_term = reference[\"properties\"][0][\"value\"]\n",
    "            go_terms.append((go_id, go_term))\n",
    "    return go_terms\n",
    "\n",
    "def get_subcellular_locations(protein_info):\n",
    "    subcell_locations = []\n",
    "    comments = protein_info[\"comments\"]\n",
    "    for comment in comments:\n",
    "        if comment[\"commentType\"] == \"SUBCELLULAR LOCATION\":\n",
    "            locations = comment[\"subcellularLocations\"]\n",
    "            for location in locations:\n",
    "                subcell_locations.append(location[\"location\"][\"value\"])\n",
    "    return subcell_locations\n",
    "\n",
    "def get_transmembrane(protein_info):\n",
    "    transmembrane = 0\n",
    "    features = protein_info[\"features\"]\n",
    "    for feature in features:\n",
    "        if feature[\"type\"] == \"Transmembrane\":\n",
    "            transmembrane = 1\n",
    "    return transmembrane\n",
    "\n",
    "def get_binding_preference(protein_info):\n",
    "    bp = []\n",
    "    keywords = protein_info[\"keywords\"]\n",
    "    for keyword in keywords:\n",
    "        kw = keyword[\"name\"]\n",
    "        if (kw == \"DNA-binding\" or kw == \"RNA-binding\"):\n",
    "            bp.append(\"DNA/RNA-binding\")\n",
    "        elif (kw == \"Nucleotide-binding\" or kw == \"Metal-binding\"):\n",
    "            bp.append(kw)\n",
    "    return bp\n",
    "\n",
    "# #--- Functions: Assign Motifs (Prosite) ---\n",
    "# def assign_motifs():\n",
    "#     global proteins\n",
    "#     with open(motif_file, 'r') as file:\n",
    "#         data = json.load(file)\n",
    "#     for id in data:\n",
    "#         proteins[id]['motifs'] = data[id]\n",
    "\n",
    "#--- Functions: Assign iFeature Data (iFeature) ---\n",
    "def assign_iFeature_data(ifeature_file, proteins):\n",
    "    with open(ifeature_file, 'r') as file:\n",
    "        features = json.load(file)\n",
    "    for feature, ids in features.items():\n",
    "        for id, value in ids.items():\n",
    "            proteins[id][feature] = value\n",
    "\n",
    "#--- Functions: Assign Protein Names* ---\n",
    "# *names exctracted from EMBL database for test proteins\n",
    "def assign_names(names_file, proteins):\n",
    "    data = []\n",
    "    with open(names_file, 'r') as file:\n",
    "        tsv_reader = csv.reader(file, delimiter='\\t')\n",
    "        next(tsv_reader)\n",
    "        for row in tsv_reader:\n",
    "            data.append(row)\n",
    "    for protein in data:\n",
    "        proteins[protein[0]]['name'] = protein[2]\n",
    "\n",
    "#--- Functions: Generate JSON Based on Dictionary\n",
    "def dict_to_json(dict, file_path):\n",
    "    with open(file_path, 'w') as json_file:\n",
    "        json.dump(dict, json_file, indent=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test Proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -- Variables: Protein Dictionary ---\n",
    "test_proteins = {}\n",
    "\n",
    "#--- Main: Generate Protein Dictionary ---\n",
    "for record in SeqIO.parse(test_fasta_file, \"fasta\"):\n",
    "    protein_id = record.id.split(\"|\")[1].strip()\n",
    "    seq = record.seq\n",
    "    a_seq = ProteinAnalysis(seq)\n",
    "\n",
    "    protein = {\n",
    "        \"seq\" : str(seq),\n",
    "        \"name\" : \"\",\n",
    "        \"length\" : len(record),\n",
    "        \"m_weight\": a_seq.molecular_weight(),\n",
    "        \"instab_index\": a_seq.instability_index(),\n",
    "        \"isoele_point\": a_seq.isoelectric_point(),\n",
    "        \"gravy\": a_seq.gravy(),\n",
    "        \"amino_count\": a_seq.count_amino_acids(),\n",
    "        \"aromaticity\": a_seq.aromaticity(),\n",
    "        \"flexibility\": a_seq.flexibility(),\n",
    "        \"sec_sruct_frac\": a_seq.secondary_structure_fraction(),\n",
    "        \"ext_coeff\": a_seq.molar_extinction_coefficient(),\n",
    "        \"go_terms\": [],\n",
    "        #\"motifs\": [],\n",
    "        \"dpc\": [],\n",
    "        # \"tpc\": [],\n",
    "        \"paac\": [],\n",
    "        \"ctdc\": [],\n",
    "        \"ctdt\": [],\n",
    "        \"ctdd\": [],\n",
    "        \"ctriad\": [],\n",
    "        \"gaac\": [],\n",
    "        \"moran\": [],\n",
    "        \"subcell_locations\": [],\n",
    "        \"transmembrane\": 0,\n",
    "        \"binding_preference\": [],\n",
    "    }\n",
    "    test_proteins[protein_id] = protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assign_annotations(test_json_file, test_proteins)\n",
    "#assign_motifs()\n",
    "assign_iFeature_data(test_ifeature_file, test_proteins)\n",
    "assign_names(test_names_file, test_proteins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_to_json(test_proteins, test_output_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Train Proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# -- Variables: Protein Dictionary ---\n",
    "train_proteins = {}\n",
    "\n",
    "#--- Main: Generate Protein Dictionary ---\n",
    "for record in SeqIO.parse(train_protein_file, \"fasta\"):\n",
    "    protein_id = record.id.split(\"|\")[1].strip()\n",
    "    seq = record.seq\n",
    "    a_seq = ProteinAnalysis(seq)\n",
    "\n",
    "    protein = {\n",
    "        \"seq\" : str(seq),\n",
    "        \"name\" : \"\",\n",
    "        \"length\" : len(record),\n",
    "        \"m_weight\": a_seq.molecular_weight(),\n",
    "        \"instab_index\": a_seq.instability_index(),\n",
    "        \"isoele_point\": a_seq.isoelectric_point(),\n",
    "        \"gravy\": a_seq.gravy(),\n",
    "        \"amino_count\": a_seq.count_amino_acids(),\n",
    "        \"aromaticity\": a_seq.aromaticity(),\n",
    "        \"flexibility\": a_seq.flexibility(),\n",
    "        \"sec_sruct_frac\": a_seq.secondary_structure_fraction(),\n",
    "        \"ext_coeff\": a_seq.molar_extinction_coefficient(),\n",
    "        \"go_terms\": [],\n",
    "        #\"motifs\": [],\n",
    "        \"dpc\": [],\n",
    "        # \"tpc\": [],\n",
    "        \"paac\": [],\n",
    "        \"ctdc\": [],\n",
    "        \"ctdd\": [],\n",
    "        \"ctdt\": [],\n",
    "        \"ctriad\": [],\n",
    "        \"gaac\": [],\n",
    "        \"moran\": [],\n",
    "        \"subcell_locations\": [],\n",
    "        \"transmembrane\": 0,\n",
    "        \"binding_preference\": [],\n",
    "    }\n",
    "    train_proteins[protein_id] = protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "assign_annotations(train_uniprot_file, train_proteins)\n",
    "#assign_motifs()\n",
    "assign_iFeature_data(train_ifeature_file, train_proteins)\n",
    "#Note: names not assigned for train proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_to_json(train_proteins, train_output_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Depreciated testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- Test: Print Proteins ---\n",
    "# for protein_id, protein_info in proteins.items():\n",
    "#     print(f\"Protein ID: {protein_id}, Name: {protein_info['name']}\")\n",
    "#     print(f\"Protein Sequence: {protein_info['seq']}\")\n",
    "#     print(f\"Protein Length: {protein_info['length']}\")\n",
    "#     print(f\"Molecular Weight: {protein_info['m_weight']}\")\n",
    "#     print(f\"Instability Index: {protein_info['instab_index']}\")\n",
    "#     print(f\"Isoelectric Point: {protein_info['isoele_point']}\")\n",
    "#     print(f\"Gravy: {protein_info['gravy']}\")\n",
    "#     print(f\"Amino Acid Count: {protein_info['amino_count']}\")\n",
    "#     print(f\"Aromaticity: {protein_info['aromaticity']}\")\n",
    "#     print(f\"Flexibility: {protein_info['flexibility']}\")\n",
    "#     print(f\"Secondary Structure Fraction: {protein_info['sec_sruct_frac']}\")\n",
    "#     print(f\"Gene Ontology Terms: {protein_info['go_terms']}\")\n",
    "    # print(f\"Motifs: {protein_info['motifs']}\")\n",
    "    # print(f\"Dipeptide Composition: {protein_info['dpc']}\")\n",
    "    # print(f\"Tripeptide Composition: {protein_info['tpc']}\")\n",
    "    # print(f\"Pseudo Amino Acid Composition: {protein_info['paac']}\")\n",
    "    # print(f\"Composition: {protein_info['ctdc']}\")\n",
    "    # print(f\"Distribution: {protein_info['ctdd']}\")\n",
    "    # print(f\"Translation: {protein_info['ctdt']}\")\n",
    "    # print(f\"Conjoint Triad: {protein_info['ctriad']}\")\n",
    "    # print(f\"Grouped Amino Acid Composition: {protein_info['gaac']}\")\n",
    "    # print(f\"Moran Autocorrelation: {protein_info['moran']}\")\n",
    "    # print(f\"Subcellular Locations: {protein_info['subcell_locations']}\")\n",
    "    # print(f\"Transmembrane?: {protein_info['transmembrane']}\")\n",
    "    # print(f\"Binding Preference: {protein_info['binding_preference']}\")\n",
    "    # print()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
